pxtlib/util.ts

```diff
@@ -1194,6 +1194,155 @@ namespace ts.pxtc.Util {
+        while (bpp < 4) {
+            if (usableBytes * bpp >= neededBytes * 8)
+                break
+            bpp++
+        }
+        let imgCapacity = (usableBytes * bpp) >> 3
+        let missing = neededBytes - imgCapacity
+        let addedLines = 0
+        let addedOffset = canvas.width * canvas.height * 4
+        if (missing > 0) {
+            const bytesPerLine = canvas.width * 3
+            addedLines = Math.ceil(missing / bytesPerLine)
+            const c2 = document.createElement("canvas")
+            c2.width = canvas.width
+            c2.height = canvas.height + addedLines
+            const ctx = c2.getContext("2d")
+            ctx.drawImage(canvas, 0, 0)
+            canvas = c2
+        }
+
+        let header = pxt.HF2.encodeU32LE([
+            imageMagic,
+            blob.length,
+            addedLines,
+            0,
+            0,
+            0,
+            0,
+            0,
+            0,
+        ])
+
+        pxt.Util.assert(header.length == imageHeaderSize)
+
+        function encode(img: Uint8ClampedArray, ptr: number, bpp: number, data: ArrayLike<number>) {
+            let shift = 0
+            let dp = 0
+            let v = data[dp++]
+            const bppMask = (1 << bpp) - 1
+            let keepGoing = true
+            while (keepGoing) {
+                let bits = (v >> shift) & bppMask
+                let left = 8 - shift
+                if (left <= bpp) {
+                    if (dp >= data.length) {
+                        if (left == 0) break
+                        else keepGoing = false
+                    }
+                    v = data[dp++]
+                    bits |= (v << left) & bppMask
+                    shift = bpp - left
+                } else {
+                    shift += bpp
+                }
+                img[ptr] = ((img[ptr] & ~bppMask) | bits) & 0xff
+                ptr++
+                if ((ptr & 3) == 3) {
+                    // set alpha to 0xff
+                    img[ptr++] = 0xff
+                }
+            }
+            return ptr
+        }
+
+        const ctx = canvas.getContext("2d")
+        const imgdat = ctx.getImageData(0, 0, canvas.width, canvas.height)
+
+        // first pixel holds bpp (LSB are written first, so we can skip what it writes in second and third pixel)
+        encode(imgdat.data, 0, 1, [bpp])
+        let ptr = 4
+        // next, the header
+        ptr = encode(imgdat.data, ptr, bpp, header)
+        pxt.Util.assert((ptr & 3) == 0)
+        if (addedLines == 0)
+            ptr = encode(imgdat.data, ptr, bpp, blob)
+        else {
+            let firstChunk = imgCapacity - header.length
+            ptr = encode(imgdat.data, ptr, bpp, blob.slice(0, firstChunk))
+            ptr = encode(imgdat.data, addedOffset, 8, blob.slice(firstChunk))
+        }
+        // set remaining alpha
+        ptr |= 3
+        while (ptr < imgdat.data.length) {
+            if (!transparent) imgdat.data[ptr] = 0xff
```

Comment:
i extracted the encode/decode functions from screenshot.ts and did not change them except for here--added a flag to leave the remaining alpha pixels as they are to preserve the blocks screenshot transparency

Reply:
I guess the `if (!transparent)` could be around the whole loop?